File: 11 Container With Most Water/containerWithMostWater.py
class Solution(object):
    def maxArea(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        if len(height) <= 1:
            return 0
        else:
            l = 0
            r = len(height) - 1
            maxArea = 0
            
            while l < r:
                while height[l] <= height[r] and l < r: # Pay attention to the boundary
                    if height[l] * (r - l) > maxArea:
                        maxArea = height[l] * (r - l)
                    l += 1
                while height[l] > height[r] and l < r: # Pay attention to the boundary
                    if height[r] * (r - l) > maxArea:
                        maxArea = height[r] * (r - l)
                    r -= 1
            return maxArea

File: 11 Container With Most Water/containerWithMostWater.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        assert(height.size() > 1);
        int l = 0;
        int r = height.size() - 1;
        int maxArea = 0;
        
        while(l < r){
            while(height[l] > height[r]){
                if(height[r] * (r - l) > maxArea){
                    maxArea = height[r] * (r - l);
                }
                r--;
            }
            while(height[l] <= height[r]){
                if(height[l] * (r - l) > maxArea){
                    maxArea = height[l] * (r - l);
                }
                l++;
            }
        }
        return maxArea;
    }
};